### PR TITLE
"rent" extra pages for app create

### PR DESCRIPTION
**Pull request by AyAggarwal**
Followed discussion from https://github.com/algorand/py-algorand-sdk/pull/196, put extraPages last in parameter order. Updated tests to explicitly populate undefined parameters for readability

### CODE REVIEW

File: src/transaction.ts
Comment (AyAggarwal) [inline src/transaction.ts:260]: all numbers in javascript are floats, should I manually check if the value is within {0,1,2,3}? or is this okay

Comment (jdtzmn) [reply src/transaction.ts:260]: Perhaps use `Number.isInteger()` method?

Comment (jasonpaulos) [reply src/transaction.ts:260]: If you add a `Number.isInteger` check in addition to this, that should be sufficient.

File: src/makeTxn.ts
Comment (jdtzmn) [inline src/makeTxn.ts:1144]: The `RenameProperties` TypeScript generic type renames properties but _passes properties as well_, so this assignment is unnecessary. Adding `'extraPages'` to the `Pick` union is sufficient.

File: tests/cucumber/steps/steps.js
Comment (jdtzmn) [inline tests/cucumber/steps/steps.js:4297]: Must these be explicitly set to `undefined`? Does the test fail when these properties are not set?

**Review by jdtzmn — APPROVED** [review]
`Number.isInteger` should be added and I had a few questions, but otherwise this looks good!

File: src/transaction.ts
Comment (AyAggarwal) [reply src/transaction.ts:260]: Ok I'll use that LOL

Comment (AyAggarwal) [reply src/transaction.ts:260]: before it was giving me an error but it works now that the test file is fixed

File: src/makeTxn.ts
Comment (AyAggarwal) [reply src/makeTxn.ts:1144]: gotcha

File: tests/cucumber/steps/steps.js
Comment (AyAggarwal) [reply tests/cucumber/steps/steps.js:4297]: In this case it may work because its from an object, but removing these parameters gives issues in other tests so I rather keep it consistent

Comment (jdtzmn) [reply tests/cucumber/steps/steps.js:4297]: I personally disagree - I think it's generally bad practice to set undefined properties on an object, but I'll wait to see what @jasonpaulos thinks.

File: src/transaction.ts
Comment (jdtzmn) [inline src/transaction.ts:260]: nit: change this to `extraPages` because that is the name of the parameter passed in.

Comment (AyAggarwal) [reply src/transaction.ts:260]: done

File: tests/cucumber/steps/steps.js
Comment (jasonpaulos) [reply tests/cucumber/steps/steps.js:4297]: In this case I'd say the `undefined` fields should be omitted since they have nothing to do with this test.

Comment (AyAggarwal) [reply tests/cucumber/steps/steps.js:4297]: cool thanks yall

**Comment by onetechnical** [comment]
You may need to adjust the chrome driver, as you have this error:

```
SessionNotCreatedError: session not created: This version of ChromeDriver only supports Chrome version 89
Current browser version is 91.0.4472.77 with binary path /usr/bin/google-chrome
    at Object.throwDecodedError (/home/travis/build/[secure]/js-[secure]-sdk/node_modules/selenium-webdriver/lib/error.js:550:15)
    at parseHttpResponse (/home/travis/build/[secure]/js-[secure]-sdk/node_modules/selenium-webdriver/lib/http.js:565:13)
    at Executor.execute (/home/travis/build/[secure]/js-[secure]-sdk/node_modules/selenium-webdriver/lib/http.js:491:26)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:97:5) {
  remoteStacktrace: '#0 0x55db65b662b9 <unknown>\n'
}
```

**Comment by jasonpaulos** [comment]
I've made #370 to update chromedriver and a few other deps, which should fix the CI issue. Hopefully it can get merged tomorrow and this PR can rebase/merge with develop.

File: src/types/transactions/base.ts
Comment (jasonpaulos) [inline src/types/transactions/base.ts:371]: Could you mention that a page is 1024 bytes, like in https://github.com/algorand/py-algorand-sdk/blob/3ca47273258ae99e168272f95b66416359a00c97/algosdk/future/transaction.py#L1161?

**Comment by jannotti** [comment]
It's a little more complicated than that now.  Because we're also combining
the limit on Approval and Clear, these pages are 2k, but shared.

**Comment by jasonpaulos** [comment]
@jannotti so what are the effects of requesting 1 extra page? An additional 1024 bytes to the approval program, or an additional 2048 bytes to the shared pool of space between the approval & clear state program?

**Comment by jannotti** [comment]
> @jannotti so what are the effects of requesting 1 extra page? An additional 1024 bytes to the approval program, or an additional 2048 bytes to the shared pool of space between the approval & clear state programs?

The latter.  You get 2 extra k, to be split as you like. So with 1 extra page you could have a nearly 4k Approval, and a tiny Clear State.

File: src/transaction.ts
Comment (jdtzmn) [inline src/transaction.ts:425]:
```suggestion
```
nit: unnecessary newline in an otherwise un-spaced section of conditionals.

File: src/transaction.ts
Comment (jdtzmn) [inline src/transaction.ts:258]: What was the issue with `Number.isInteger`?

Comment (AyAggarwal) [reply src/transaction.ts:258]: It somehow causes all the tests to fail with extra pages parameter to fail. Odd because they are most definitely integers

Comment (AyAggarwal) [reply src/transaction.ts:258]: It's because I didn't check for undefined first 🤦‍♂️

Comment (jasonpaulos) [reply src/transaction.ts:258]:
```suggestion
    if (txn.extraPages != null && (!Number.isInteger(txn.extraPages) || txn.extraPages < 0 || txn.extraPages > 3))
```

**Review by jasonpaulos — CHANGES_REQUESTED** [review]
Good except for the missing `Number.isInteger` check. I will approve immediately once that's added.

**Comment by jasonpaulos** [comment]
Apologies, it seems my suggested change did not satisfy prettier or eslint ☹️

**Comment by AyAggarwal** [comment]
> Apologies, it seems my suggested change did not satisfy prettier or eslint ☹️

That's okay I force pushed my local version

File: src/transaction.ts
Comment (jdtzmn) [inline src/transaction.ts:260]: I'm not sure this null check is necessary

**Review by jdtzmn — CHANGES_REQUESTED** [review]

File: src/transaction.ts
Comment (AyAggarwal) [reply src/transaction.ts]: in my original commit it checks for undefined, which is necessary because extrapages is only defined in create tests. otherwise the error gets triggered because undefined is not an integer

Comment (jasonpaulos) [reply src/transaction.ts]: Right, I suppose `Number.isInteger` would return false for null, rending the null check a little redundant.

Comment (AyAggarwal) [reply src/transaction.ts]: its not redundant because we don't want it to return false for undefined. this test should only apply to cases where the parameter was passed in

Comment (jasonpaulos) [reply src/transaction.ts]: Great point!

**Comment by AyAggarwal** [comment]
BTW I have #368 merged in here right now

**Pull request merged**
